payment/controllers.go

```diff
@@ -30,6 +31,126 @@ func Router(service *Service) chi.Router {
+func CreateKey(service *Service) handlers.AppHandler {
+	return handlers.AppHandler(func(w http.ResponseWriter, r *http.Request) *handlers.AppError {
+		var req CreateKeyRequest
+		err := requestutils.ReadJSON(r.Body, &req)
+		if err != nil {
+			// FIXME Ask Ben what he would like us to do here instead of wrapError
+			return handlers.WrapError(err, "Error in request body", http.StatusBadRequest)
+		}
+
+		_, err = govalidator.ValidateStruct(req)
+		if err != nil {
+			return handlers.WrapValidationError(err)
+		}
+
+		encrypted, nonce := GenerateSecret()
+		// var Key Key
+		key, err := service.datastore.CreateKey(req.Merchant, encrypted, nonce)
+		if err != nil {
+			return handlers.WrapError(err, "Error create api keys", http.StatusInternalServerError)
+		}
+
+		w.WriteHeader(http.StatusOK)
+
+		if err := json.NewEncoder(w).Encode(key); err != nil {
+			return handlers.WrapError(err, "Error encoding the keys JSON", http.StatusInternalServerError)
+		}
+
+		return nil
+	})
+}
+
+// DeleteKey deletes a key
+func DeleteKey(service *Service) handlers.AppHandler {
+	return handlers.AppHandler(func(w http.ResponseWriter, r *http.Request) *handlers.AppError {
+		reqID := chi.URLParam(r, "id")
+		if reqID == "" || !govalidator.IsUUIDv4(reqID) {
+			return handlers.ValidationError(
+				"Error validating request url parameter",
+				map[string]interface{}{
+					"id": "id must be a uuidv4",
+				},
+			)
+		}
+
+		id := uuid.Must(uuid.FromString(reqID))
+
+		var req DeleteKeyRequest
+		err := requestutils.ReadJSON(r.Body, &req)
+		if err != nil {
+			// FIXME Ask Ben what he would like us to do here instead of wrapError
+			return handlers.WrapError(err, "Error in request body", http.StatusBadRequest)
+		}
+
+		_, err = govalidator.ValidateStruct(req)
+		if err != nil {
+			return handlers.WrapValidationError(err)
+		}
+
+		key, err := service.datastore.DeleteKey(id, req.DelaySeconds)
+		if err != nil {
+			return handlers.WrapError(err, "Error deleting the keys", http.StatusInternalServerError)
+		}
+
+		w.WriteHeader(http.StatusOK)
+
+		if err := json.NewEncoder(w).Encode(key); err != nil {
+			return handlers.WrapError(err, "Error encoding the keys JSON", http.StatusInternalServerError)
+		}
+
+		return nil
+	})
+}
+
+// GetKeys FIXME
+func GetKeys(service *Service) handlers.AppHandler {
+	return handlers.AppHandler(func(w http.ResponseWriter, r *http.Request) *handlers.AppError {
+		reqID := chi.URLParam(r, "merchant")
+		expired := r.URL.Query().Get("expired")
+		showExpired := expired == "true"
+
+		var keys *[]Key
+		keys, err := service.datastore.GetKeys(reqID, showExpired)
+		if err != nil {
+			return handlers.WrapError(err, "Error deleting the keys", http.StatusInternalServerError)
+		}
+
+		w.WriteHeader(http.StatusOK)
+
+		if err := json.NewEncoder(w).Encode(keys); err != nil {
+			return handlers.WrapError(err, "Error encoding the keys JSON", http.StatusInternalServerError)
+		}
+
+		return nil
```

Comment:
If there is a problem encoding the JSON you will have already sent the 200 Response Header, and then there will be a panic trying to resend a new 500 Response Header.  You can only WriteHeader once.  This might be cool if we could abstract this a bit into a `RenderContent` function, which will use the request's `Accept` header to figure out how to render the response (JSON, XML, etc).  Would help reusability.

Reply:
Here we go! 

https://github.com/brave-intl/bat-go/pull/285/commits/e213768937e3a0d31d87fdd83260060d058cce3a